src/server/transaction.cc

```diff
@@ -116,6 +116,40 @@ uint32_t Transaction::PhasedBarrier::DEBUG_Count() const {
   return count_.load(memory_order_relaxed);
 }
 
+bool Transaction::SingleClaimBarrier::IsClaimed() const {
+  return claimed_.load(memory_order_relaxed);
+}
+
+bool Transaction::SingleClaimBarrier::TryClaim() {
+  return !claimed_.exchange(true, memory_order_relaxed);  // false means first means success
+}
+
+void Transaction::SingleClaimBarrier::Release() {
+  DCHECK(claimed_.load(memory_order_relaxed));
+  released_.store(true, memory_order_relaxed);
+  ec_.notify();  // release
+}
+
+cv_status Transaction::SingleClaimBarrier::Wait(time_point tp) {
+  auto cb = [this] { return released_.load(memory_order_acquire); };
+
+  if (tp != time_point::max()) {
+    cv_status status = ec_.await_until(cb, tp);
+
+    if (status == cv_status::no_timeout)  // We finished without a timeout due to a release
+      return cv_status::no_timeout;
+
+    if (!TryClaim())  // If we can't claim the barrier after a timeout, someone is modifying us
+      return Wait(time_point::max());  // wait for the modification to finish
+
+    Release();  // Otherwise release immediately and finish
+    return cv_status::timeout;
```

Comment:
why do we need to release in this case? is there any well defined state for this barrier once Wait finishes? 

```suggestion
    if (TryClaim())  // If we claimed the barrier, no one will be able to claim it anymore
      return cv_status::timeout;
```

Reply:
Not sure, just to be consistent, I don't want it formally to be hanging in an undefined state

Reply:
in that case, please still reverse the condition and avoid the recursive call by continuing out of the block.